rfcs/2020-11-01-remap-metrics-support.md

```diff
@@ -0,0 +1,199 @@
+# RFC 3550 - 2020-11-01 - Support metrics with the new 'remap' transform
+
+It would be useful to add the ability for the Remap language to work with metric
+data as well as with log data.
+
+## Scope
+
+The aim of the Remap language is to remain simple and yet provide the power to
+allow the user to replace most of their tranform configuration with a simple
+Remap script. Any complexity beyond a certain level is possible, but requires
+the `Lua` or `Wasm` transforms.
+
+Given this, the scope of this RFC is not to enhance the Remap language so it
+can work comprehensively with metric data. The main aim will be to allow the
+user to make simple changes to the metadata surrounding the metric.
+
+
+## Motivation
+
+Currently there are some simple transforms that allow manipulating the metadata
+of metric data - `add_tags`, `remove_tags`, `rename_tags`. These transforms
+are useful, but limited. They cannot take into account any additional data around
+the metric to determine how to process the tags.
+
+
+## Doc-level Proposal
+
+The Remap language can work with the metadata around a metric.
+
+When the event the Remap function is working with, the metric metadata can be 
+accessed through a number of set fields in a similar way to how the language 
+works with log events.
+
+The following fields are available:
+
+- `.name` - The name of the metric.
+- `.namespace` - The namespace for the metric.
+- `.timestamp` - Timestamp of the metric.
+- `.kind` - The kind of the metric - *Incremental* or *Absolute*
+- `.tags` - A map containing the tags set for the metric.
+ 
+There are strict limitations to how this data can be represented. 
+
+Setting `name` or `namespace` to anything other than a string will be an 
+error. `.timestamp` has to be a timestamp. `tags` is a map of string keys to 
+string values. No nesting is allowed. 
```

Comment:
It's good to think through the technical implications here.

I'd vote to add a **Internal Proposal** section to the RFC similar to other RFCs that provide technical changes to Vector.

The reason for this is that I'd like us to find the right solution towards providing capabilities such as the ones written above (e.g. providing limitations on field assignments), without creating too much technical debt within the language.

That is to say, I see several routes we could take here:

1. Run the Remap language in a special "metrics" mode, which would use a custom parser and grammar. This is fairly straightforward and gives us compile-time checking out of the box, but is also the most decoupled and provides us with the least amount of reusability within the language between metrics and logs.

2. Have field access / mutation check at runtime. This is the simplest solution, since we only need to update `impl remap::Object for vector::Event` in Vector itself, and have it check if we're dealing with a metric event, and check field names and value types accordingly. The downside of this is that a user doesn't know their script is invalid until it is running, i.e. we can't do any compile-time checking.

3. Extend the existing compile-time checks we've been introducing (#4902, #5008, #5015) to also provide ways to check which fields are being queried/assigned and what values are being assigned to them. This is the most extensible solution that also allows us to re-use this logic for regular logs (for when we introduce log schemas), but it's also the most work, and the most hand-wavy right now, as I haven't thought through the technical implications yet.

Of course, we could always revert to your initial proposal of not using paths at all, and only manipulating metrics through functions, but to me, that makes the whole experience feel second-tier in Remap itself, and it makes metric support feel like more of a tacked-on solution (and it's not entirely clear yet if that would solve all our runtime vs compile-time error needs).

Reply:
Thinking through number (3) a bit more:

Querying a metric through paths is basically **querying an event with a strict schema**.

If we were to introduce the concept of _`Object` schemas_ to the Remap language, we'd be able to:

1. Add a custom `metrics` schema which would enforce the path invariants described in this RFC.
2. In the future, we could use schema metadata (#4599) to enforce more strict type checking at compile-time for logs (in other words, right now, every field in a log is considered to be of type `Any` before you constraint fields through functions (e.g. `to_string(.foo)`). After this change, we can know at compile-time if a field is `Any`, or `String`, or any other kind).

The idea is basically an extension of #4902, with an event schema for metrics that looks something like this (ignoring naming):

```rust
use remap::event::Constraint::*;

ObjectSchema {
	allow_accessing_undefined_paths: false,
	schema: map![
		".name": FieldDef { constraint: Exact(String), mutable: true },
		// ...
		".tags": FieldDef { path: ".tags", constraint: Exact(Map), mutable: true },
	],
}
```

Based on this schema, we can then check (at compile-time) if field querying/assignments are valid. You'd be able to do:

```rust
.name = "foo" 			// ok
.name = to_string(1) 	// ok
.name = 1 				// compile error: field ".name" expected string, got integer
.foo = true 			// compile error: unable to assign to unknown field ".foo"
```

We can start by applying this schema at compile-time, and leaving it for the duration of the program, but one of the future proposals made in this RFC is to convert a metric to a log, which could look something like this:

```rust
.name = "foo"

// .foo = true
// ^ this would result in a compile-time error

to_log()

.foo = true
// ^ this now works, since the schema was changed by `to_log()` to allow any paths with any values
```

But I'd consider that out of scope for this RFC, as this brings with it its own set of complexities.